### PR TITLE
[Internal] Refine error message for tool meta generator.

### PR DESCRIPTION
**Pull request by codingrabbitt1**
# Description

Refine error message for tool meta generator.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 1]
```diff
@@ -53,11 +57,13 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
 
     for input in inputs:
         if input in reserved_keys_to_raise:
-            msg = (
-                f"Parsing jinja got exception: Variable name {input} is reserved by {tool_type.value} tool."
-                + " Please change another name."
+            raise ReservedVariableCannotBeUsed(
+                message_format="Generate tool meta failed for PROMPT tool. "
+                "Jinja parsing failed due to reserved variable conflict. "
+                "Variable name {key} is reserved by {tool_type} tool, please change another name.",
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:63]: How about:
```Jinja parsing failed: Variable name '{key}' is reserved name by {tool_type} tools, please change to another name.```

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 2]
```diff
@@ -119,7 +125,12 @@ def _parse_tool_from_function(f):
     try:
         inputs, _, _ = function_to_interface(f)
     except Exception as e:
-        raise BadFunctionInterface(f"Failed to parse interface for tool {f.__name__}, reason: {e}") from e
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
+        raise BadFunctionInterface(
+            message_format="Parse interface for tool {tool_name} failed: {error_type_and_message}",
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:130]: Add quota to tool_name:

`{tool_name}` -> `'{tool_name}'`

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 3]
```diff
@@ -151,13 +162,22 @@ def load_python_module_from_file(src_file: Path):
...
     m = importlib.util.module_from_spec(spec)
     try:
         spec.loader.exec_module(m)
     except Exception as e:
         # TODO: add stacktrace to additional info
-        raise PythonLoadError(f"Failed to load python module from file '{src_file}', reason: {e}.") from e
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
+        raise PythonLoadError(
+            message_format="Failed to load python module from file '{src_file}'. "
+            "Error details: {error_type_and_message}",
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:177]: Let's use this style:
`Failed to load python module from file '{src_file}': {error_type_and_message}`

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 4]
```diff
@@ -218,18 +251,23 @@ def generate_tool_meta_dict_by_file(path: str, tool_type: ToolType):
...
     try:
         content = file.read_text()
     except Exception as e:
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
         raise MetaFileReadError(
-            message_format=("Reading meta file failed, path '{file_path}', type '{tool_type}', exception {exception}."),
-            file_path=str(file),
+            message_format=(
+                "Generate tool meta failed for {tool_type} tool. "
+                "Read meta file '{file_path}' failed. Error details: {error_type_and_message}"
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:266]: `Read meta file '{file_path}' failed: {error_type_and_message}`

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 5]
```diff
@@ -37,8 +37,12 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
     try:
         inputs = get_inputs_for_prompt_template(content)
     except Exception as e:
-        msg = f"Parsing jinja got exception: {e}"
-        raise JinjaParsingError(msg) from e
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
```
Comment (ChanningWang) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:40]: ask zhidong how to get error type and message from e

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 4 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 3 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 2 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 1 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 6]
```diff
@@ -37,8 +37,12 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
...
+        raise JinjaParsingError(
+            message_format="Generate tool meta failed for PROMPT tool. "
+            "Jinja parsing failed: {error_type_and_message}",
```
[the remark this answers is not in the document]
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 7]
```diff
@@ -37,8 +37,12 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
     try:
         inputs = get_inputs_for_prompt_template(content)
     except Exception as e:
-        msg = f"Parsing jinja got exception: {e}"
-        raise JinjaParsingError(msg) from e
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
+        raise JinjaParsingError(
+            message_format="Generate tool meta failed for PROMPT tool. "
```
[the remark this answers is not in the document]
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Not necessary, all the errors thrown from tool_meta_generator.py will be wrapped as ToolLoadError, which will contain the node name.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 8]
```diff
@@ -53,11 +59,14 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
 
     for input in inputs:
         if input in reserved_keys_to_raise:
-            msg = (
-                f"Parsing jinja got exception: Variable name {input} is reserved by {tool_type.value} tool."
-                + " Please change another name."
+            raise ReservedVariableCannotBeUsed(
+                message_format=(
+                    "Generate tool meta failed for {tool_type} tool. Jinja parsing failed: "
+                    "Variable name '{key}' is reserved name by {tool_type} tools, please change to another name."
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:65]: is reserved name -> is a reserved name

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 9]
```diff
@@ -170,17 +192,34 @@ def load_python_module(content, source=None):
...
+        ) from e
 
 
 def collect_tool_function_in_module(m):
     tools = collect_tool_functions_in_module(m)
     if len(tools) == 0:
-        raise NoToolDefined("No tool found in the python script.")
+        raise NoToolDefined(
+            message_format=(
+                "The number of tools defined is illegal. "
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:207]: I feel like we can remove this line. It makes the error message long and hard to read.

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 10]
```diff
@@ -240,14 +283,22 @@ def generate_tool_meta_dict_by_file(path: str, tool_type: ToolType):
     elif tool_type == ToolType.PROMPT:
         return generate_prompt_meta_dict(name, content, prompt_only=True, source=path)
     else:
-        raise ValueError(f"Unsupported tool type {tool_type}.")
+        raise NotSupported(
+            message_format=(
+                "Generate tool meta failed. "
+                "The type '{tool_type}' is currently unsupported. "
+                "Please choose from available types: {supported_tool_type} and try again."
```
Comment (zzn2) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:290]: supported_tool_type -> supported_tool_types

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 8 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 9 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 10 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py]: Done.

**Review by zzn2 — DISMISSED** [review]

File: src/promptflow/promptflow/_core/tool_meta_generator.py [hunk 11]
```diff
@@ -37,8 +37,14 @@ def generate_prompt_tool(name, content, prompt_only=False, source=None):
     try:
         inputs = get_inputs_for_prompt_template(content)
     except Exception as e:
-        msg = f"Parsing jinja got exception: {e}"
-        raise JinjaParsingError(msg) from e
+        error_type_and_message = f"({e.__class__.__name__}) {e}"
+        raise JinjaParsingError(
+            message_format=(
+                "Generate tool meta failed for {tool_type} tool. Jinja parsing failed: {error_type_and_message}"
```
Comment (huaiyan) [inline src/promptflow/promptflow/_core/tool_meta_generator.py:43]: Also add name info in error message?
Or that may be hard to find which node needs to fix.

Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py:43]: Not necessary, all the errors thrown from tool_meta_generator.py will be wrapped as ToolLoadError, which will contain the node name.

Comment (huaiyan) [reply src/promptflow/promptflow/_core/tool_meta_generator.py:43]: Checked about call hierarchy about `generate_prompt_tool`, seems that's not true..

**Review by huaiyan — APPROVED** [review]

File: src/promptflow/promptflow/_core/tool_meta_generator.py (hunk 11 above)
Comment (codingrabbitt1) [reply src/promptflow/promptflow/_core/tool_meta_generator.py:43]: For flow run scenario, the error thrown will be wrapped as ToolLoadError. For the other scenarios, let‘s deal with them later.

**Review by huaiyan — APPROVED** [review]

**Review by zzn2 — APPROVED** [review]

**Pull request merged**